barretenberg/cpp/pil/vm2/alu.pil

```diff
@@ -201,27 +204,101 @@ sel_mul_u128 * (1 - sel_tag_err)
+
+// This is integer division, so we disallow FF (use tag err here?)
+sel_op_div * sel_is_ff = 0;
+
+// We need to show that remainder < b (remainder stored in helper1)
+// TODO(MW): Reuse #[INT_GT] by setting lt_ops_input_a = b, lt_ops_input_b = helper1, lt_ops_result_c = 1
+#[GT_DIV_REMAINDER]
+sel_op_div { ib, helper1, sel_op_div } in gt.sel { gt.input_a, gt.input_b, gt.res };
+
+// DIV - non u128
+
+// a / b = c + remainder ==> show a - remainder = b * c
+#[ALU_DIV_NON_U128]
+sel_op_div * IS_NOT_U128 * (1 - sel_tag_err)
+    * (
+        ib * ic
+        - (ia - helper1)
+        // TODO: do we need some carrying here e.g. (max_value + 1) * n?
+    ) = 0;
+
+// DIV - u128
+
+pol commit sel_div_u128;
+// sel_op_div & sel_is_u128:
+sel_div_u128 - sel_is_u128 * sel_op_div = 0;
+
+// Using this to switch on the range checks for decomposed u128 integers
+pol commit sel_mul_div_u128;
+// sel_mul_u128 || sel_div_u128
+sel_mul_div_u128 - (sel_mul_u128 + sel_div_u128) = 0;
+
+// We share the logic from MUL, since here we also must show a product evaluates to a result:
+//  c * b = a - remainder := res
+
+// Taken from vm1:
+// We express c, b in 64-bit slices: c = c_l + c_h * 2^64
+//                                   b = b_l + b_h * 2^64
+// => c * b = c_l * b_l + (c_h * b_l + c_l * b_h) * 2^64 + (c_h * b_h) * 2^128 = res_hi_full * 2^128 + res
+// => the 'top bits' are given by (res_hi_full - (c_h * b_h)) * 2^128
+// We can show for a 64 bit res_hi = res_hi_full - (c_h * b_h) % 2^64 that:
+// c_l * b_l + (c_h * b_l + c_l * b_h) * 2^64 = res_hi * 2^128 + res
+// Equivalently (cf = 0 if c_h & b_h = 0):
+// c * b_l + c_l * b_h * 2^64 = (cf * 2^64 + res_hi) * 2^128 + res
+// => no need for c_h in final relation
+
+// Reuse columns b_lo, b_hi from MUL above to represent b (see B_DECOMPOSITION)
+// Reuse columns a_lo, a_hi from MUL above to represent c (see A_DECOMPOSITION)
+// Reuse column c_hi from MUL above to represent res_hi (see RANGE_CHECK_MUL_U128_C_HI)
+
+// TODO(MW): I'm unsure whether this protects against a - remainder underflowing, but I think it might
+// because we have from the below relation:
+// a - remainder
+//  + res_hi * 2^128 + (cf ? 2^192 : 0) <-- res_hi is 64 bits => cannot be over 2^192
+//  = c * b_lo + c_lo * b_hi * TWO_POW_64 <-- all are 64 bits => cannot be over 2^192
+// where a, b are tagged to be under 2^128, and we check r < b < 2^128. But I'm not 100% sure (brain hurt)
+
+// TODO(MW): Remove cf here - we cannot have a valid b, c s.t. c * b = a - r = 128 bits with both b and c being over 64 bits
+// => cf is always 0 and below relation should protect against underflows (see above point?)
+
+#[ALU_DIV_U128]
+sel_div_u128 * (1 - sel_tag_err)
+    * (
+        ic * b_lo + a_lo * b_hi * TWO_POW_64            // c * b without the hi bits
+        - (ia - helper1)                              // a - r
+        - (max_value + 1) * (cf * TWO_POW_64 + c_hi)    // res_hi * 2^128 + (cf ? 2^192 : 0)
+    ) = 0;
+
```

Comment:
Could use some more eyes on this! It's pretty gross stuff so I'm happy to explain over a call next week though.